### PR TITLE
New Date implementation

### PR DESCRIPTION
**Pull request by jaesivsm**
* less internal values for the Date class,   introducing the use of `__slots__`
* introducing Accuracy
* allowing comparison across accuracies
* cleaning some unused Task method related to Dates
* proper timezone handling

### CODE REVIEW

**Review by Neui — CHANGES_REQUESTED** [review]
Nice rewrite. I reviewed it mid-night, so I want to take a better look later and try it out. What I am unsure about/want to look later:

* Sorting/comparing with fuzzy dates
* I think I had something else in mind but forgot about it.

File: GTG/core/dates.py [hunk 1]
```diff
@@ -41,7 +43,6 @@
 }
 
 STRINGS = {
-    # Translators: Used for display
```
[the remark this answers is not in the document]
Comment (jaesivsm) [reply GTG/core/dates.py]: Indeed, missed that in the rewrite, thanks !

File: GTG/core/dates.py [hunk 2]
```diff
@@ -51,274 +52,266 @@
...
+         SOON: lambda: date.today() + timedelta(days=15),
+         SOMEDAY: lambda: date.today() + timedelta(days=365),
+         NODATE: lambda: date.max - timedelta(days=1)}
 
 
-def convert_datetime_to_date(aday):
-    """ Convert python's datetime to date.
-    Strip unusable time information. """
-    return datetime.date(aday.year, aday.month, aday.day)
+class Accuracy(Enum):  # from less accurate from the most
```
[the remark this answers is not in the document]
Comment (jaesivsm) [reply GTG/core/dates.py]: Indeed a more thorough comment was needed here. Fixed !

File: GTG/core/dates.py [hunk 3]
```diff
@@ -51,274 +52,266 @@
...
+                        if accuracy is Accuracy.date:
+                            self.dt_value = self.dt_value.date()
+                        break
+                    except ValueError:
+                        pass
+            if self.dt_value is None:
+                try:  # it must be a fuzzy date
+                    self.dt_value = LOOKUP[str(value).lower()]
+                except KeyError:
+                    pass
```
[the remark this answers is not in the document]
Comment (jaesivsm) [reply GTG/core/dates.py]: Way cleaner indeed. Done!

File: GTG/core/dates.py [hunk 4]
```diff
@@ -24,13 +24,15 @@
...
 import locale
+from datetime import date, datetime, timedelta, timezone
+from enum import Enum
+from gettext import gettext as _
+from gettext import ngettext
 
-from gettext import gettext as _, ngettext
-
-__all__ = 'Date',
+__all__ = ['Date']
```
[the remark this answers is not in the document]
Comment (jaesivsm) [reply GTG/core/dates.py]: I think at first I just wanted to correct the "tuple by trailing comma" bad pattern, but indeed, Accuracy is needed here. Thanks !

File: GTG/core/dates.py [hunk 5]
```diff
@@ -51,274 +52,266 @@
...
+                    pass
+            if self.dt_value is None:
+                raise ValueError(f"Unknown value for date: '{value}'")
+        elif value in LOOKUP:
+            self.dt_value = LOOKUP[value]
         else:
             raise ValueError(f"Unknown value for date: '{value}'")
 
+    @property
+    def accuracy(self):
```
[the remark this answers is not in the document]
Comment (jaesivsm) [reply GTG/core/dates.py:160]: I also thought about that but `cached_property` implies not using `__slots__` (or to add `__dict__` to `__slots__`, which, in my opinion, kinda defeat the purpose).
Even if not for `cached_property` I could cache the accuracy value (to be compatible with python < 3.8). 
But I don't think the method is worth it. `def accuracy` first does class comparison which is fairly quick and then check on `datetime.datetime` or `datetime.date` attributes (which use `__slots__`, also fairly quick). I'm not convinced the gain brought by caching the result would justify the overhead.

**Comment by diegogangl** [comment]
Hey, thanks. I was thinking `Date` needed a thorough rethinking while working on the core rewrite branch, but had no time for that :)
I'll take a closer look over the weekend, since I'm busy with life issues.

File: GTG/core/dates.py [hunk 6]
```diff
@@ -51,274 +52,266 @@
...
-        if self._fuzzy is not None:
-            return STRINGS[self._fuzzy]
-        else:
-            return self._real_date.isoformat()
+        if self.accuracy is Accuracy.fuzzy:
+            return STRINGS[self.dt_value]
+        return self.dt_value.isoformat()
+
+    @property
+    def date_str(self):
```
[the remark this answers is not in the document]
Comment (jaesivsm) [reply GTG/core/dates.py]: Good point, I made the changes.

File: GTG/core/dates.py [hunk 7]
```diff
@@ -51,274 +52,266 @@
...
-        if self._fuzzy == NODATE:
+        if self.dt_value == NODATE:
             return None
-        else:
-            return (self.date() - datetime.date.today()).days
+        return (self.dt_by_accuracy(Accuracy.date) - date.today()).days
 
-    @classmethod
-    def today(cls):
+    def is_fuzzy(self):
```
[the remark this answers is not in the document]
Comment (jaesivsm) [reply GTG/core/dates.py]: Indeed, done !

**Comment by diegogangl** [comment]
Did a quick review, and it looks good! Some nitpicks here and there. 
I would also squash the commits, unless there is some change you feel must be in a separate commit in the history

**Review by diegogangl — CHANGES_REQUESTED** [review]

**Comment by jaesivsm** [comment]
Thank you guys for your reviews !

I worked on that while working on my caldav branch, and thought it'd be better off in its own PR. I think it'd have its place in the next release but in its own commit.

> I would also squash the commits, unless there is some change you feel must be in a separate commit in the history
Of course ! I only didn't rebase them myself just so application of suggested fixes would be noticeable. But none of that has its place in the master branch :)

**Comment by jaesivsm** [comment]
While rebasing my branch in one commit, I observed something strange, `Date.now()`, `Date('now')`, `Date.parse('now')` and all the variations around that pattern returns a fuzzy date. Which brings me to the following two points : 
 * `NOW` ain't fuzzy x) it's just now, so it's conceptually weird to mark it as fuzzy
 * Fuzzy values persisting as so even through writing in an xml file is weird. For example if I say due date is "NOW" and comes back tomorrow, it won't be "Yesterday" then, it'll still be NOW.

So in my last commit (kept it separated so it can be reverted) I changed that behaviour and made all "NOW" return straight forward `datetime.now()`. 

Hoping that's wise ^^

**Comment by Neui** [comment]
IMHO while questionable whenever the special "now" date should be stay fuzzy or not (appears while setting the due date) and whenever it has a valid use case, I think that should be discussed as an Issue and not done in this PR (unless of course it has been decided).

**Comment by jaesivsm** [comment]
@Neui that's a very good point (and why I kept the commit apart until now)
I moved it out of this PR and gonna open a new one to discuss opportunity and implementation

File: GTG/core/dates.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Neui) [reply GTG/core/dates.py:154]: Yeah I'd thought it could be implemented like that. I looked at the code outside the diff (where it looks even better) for very rough benchmark (was like 1 or 2 seconds slower at startup, for a sample size of 1 however), I wonder if the property could directly be calculated in the constructor, adding `accuracy` to `__slots__`. They're not big methods, so it should still be nice to look at in the code.

[the remark this answers is not in the document]
Comment (jaesivsm) [reply GTG/core/dates.py:154]: I implemented caching for testing purpose, the diff seems to be from 2.5us with the caching to 3.3us without it. Here is the code to test it with.
```python
from GTG.core.dates import Date
from datetime import date
import timeit
def test():
    d = Date(date.today())
    for i in range(10):
        a = d.accuracy
timeit.timeit(test, number=100000)
```
The gain seems small enough to me not to be tempted to complexify / uglify the code, does it ?

**Review by Neui — CHANGES_REQUESTED** [review]
Looks OK in general, need some work. I was sometimes a bit picky too since I was going to request changes anyway.

After fixing that one xml_str thing, it at least started up and looked fine. Would need some more basic testing for some of the features (recurring tasks, etc.).

**Comment by Neui** [comment]
I also tried to benchmark it using [hyperfine](https://github.com/sharkdp/hyperfine/) this time:

```
$ hyperfine --warmup 3 --prepare "git checkout upstream/master" "./launch.sh -s alot -t master" --prepare "git checkout new-date-inline" "./launch.sh -s alot -t new-date-inline" --prepare "git checkout new-date" "./launch.sh -s alot -t new-date"
Benchmark #1: ./launch.sh -s alot -t master
  Time (mean ± σ):     19.446 s ±  0.114 s    [User: 24.245 s, System: 1.497 s]
  Range (min … max):   19.227 s … 19.594 s    10 runs

Benchmark #2: ./launch.sh -s alot -t new-date-inline
  Time (mean ± σ):     19.488 s ±  0.112 s    [User: 24.174 s, System: 1.556 s]
  Range (min … max):   19.289 s … 19.676 s    10 runs

Benchmark #3: ./launch.sh -s alot -t new-date
  Time (mean ± σ):     19.821 s ±  0.185 s    [User: 24.565 s, System: 1.500 s]
  Range (min … max):   19.558 s … 20.117 s    10 runs

Summary
  './launch.sh -s alot -t master' ran
    1.00 ± 0.01 times faster than './launch.sh -s alot -t new-date-inline'
    1.02 ± 0.01 times faster than './launch.sh -s alot -t new-date'
```

`new-date` is this PR with the xml_str fix, `new-date-inline` is where I calculated the `accuracy` directly in `__init__`, `master` is upstream master.
I placed `app.quit()` at the very end of `MainWindow.__init__()` (in `GTG/gtk/browser/main_window.py`), so it would do all its stuff and then exit. The "alot" dataset is a personal one, containing quite a lot of tasks.

In other words, rather very little impact.

File: GTG/core/dates.py [hunk 8]
```diff
@@ -24,301 +24,305 @@
...
+class Date:
     """A date class that supports fuzzy dates.
 
-    Date supports all the methods of the standard datetime.date class. A Date
+    Date supports all the methods of the standard date class. A Date
     can be constructed with:
       - the fuzzy strings 'now', 'soon', '' (no date, default), or 'someday'
       - a string containing an ISO format date: YYYY-MM-DD, or
-      - a datetime.date or Date instance, or
+      - a date or Date instance, or
```
[the remark this answers is not in the document]
Comment (jaesivsm) [reply GTG/core/dates.py:107]: well not in the context, since import has been changed from `import datetime` to `from datetime import date, datetime [...]` but OK

File: GTG/core/xml.py [hunk 9]
```diff
@@ -86,7 +72,7 @@ def task_from_element(task, element: etree.Element):
     try:
         recurring_updated_date = recurring.find('updated_date').text
         if recurring_updated_date:
-            task.set_recurring_updated_date(datetime.fromisoformat(recurring_updated_date))
+            task.set_recurring_updated_date(Date(recurring_updated_date))
```
[the remark this answers is not in the document]
Comment (jaesivsm) [reply GTG/core/xml.py:75]: fuzzy date do compare with other dates object and have a valid representation as a datetime object, so I don't see the problem. We could force their accuracy in the Task.setter or extract the parser out of Date class to use it here. Or even drop the Date class from that kind of internal value altogether. 
But in itself, it doesn't seem absurd to unify all date treatment through the Date class.

[the remark this answers is not in the document]
Comment (Neui) [reply GTG/core/xml.py:75]: I don't mean that using GTG Date is bad, but rather files can contain "weird data" that GTG Date parses, but does not really make sense (think of having the updated time being a fuzzy date "soon" - it doesn't really make any sense in the context, even if you can just "convert" it).

Additionally, you can construct files that GTG would accept on one machine but not on a different machine by relying on the language (like you can write "bald" there, and GTG on my german machine would accept the file, but not on a french one). Admittedly, you need to manually modify the files, but I think you should reduce such surprises and be relatively strict.

File: tests/tools/test_dates.py [hunk 10]
```diff
@@ -55,26 +55,27 @@ def test_parses_today_as_today(self):
...
         """ Print fuzzy dates in localized version """
-        self.assertEqual(str(Date.parse("soon")), _("soon"))
-        self.assertEqual(str(Date.parse("later")), _("someday"))
-        self.assertEqual(str(Date.parse("someday")), _("someday"))
-        self.assertEqual(str(Date.parse("")), "")
+        self.assertEqual(Date("now").localized_str, _("now"))
+        self.assertEqual(Date("soon").localized_str, _("soon"))
+        self.assertEqual(Date("later").localized_str, _("someday"))
+        self.assertEqual(Date("someday").localized_str, _("someday"))
+        self.assertEqual(str(Date("")), "")
```
Comment (Neui) [inline tests/tools/test_dates.py:78]: I think you want to use `.localized_str` here anyway to "match" the old behaviour.

File: GTG/core/xml.py [hunk 11]
```diff
@@ -37,40 +37,26 @@ def task_from_element(task, element: etree.Element):
...
-        task.set_start_date(start)
-
+    for key, set_date in (('modified', task.set_modified),
+                          ('added', task.set_added_date),
+                          ('due', task.set_due_date),
+                          ('done', task.set_closed_date),
+                          ('start', task.set_start_date)):
+        value = dates.find(key)
+        if value is not None and value.text:
+            set_date(Date(value.text))
```
[the remark this answers is not in the document]
Comment (jaesivsm) [reply GTG/core/xml.py:51]: I don't change the behavior that much (only order of fuzzy / non fuzzy value is kinda altered, but i can change that), I mainly do some refactoring. If a format changes occur I'm not sure it should be handled in that PR. We'd be better doing that in a separate one.

**Review by diegogangl — APPROVED** [review]

**Comment by diegogangl** [comment]
Sorry I've been MIA, I'll try to get this reviewed soon

File: GTG/core/xml.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (diegogangl) [reply GTG/core/xml.py:75]: Don't think this would be an issue, since this isn't a user-facing variable it won't be set to fuzzy dates

**Comment by diegogangl** [comment]
LGTM! There are some conflicts, probably from the PEP8 PR (sorry). I'll merge right away once those are solved.
